add fragments to return

bcoca opened the pull request:
also minor refactor in getting docstrings from import
  making it easier to reuse
  
  Fixes #67047

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Docs Pull Request
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
ansible-doc

felixfontein commented on lib/ansible/utils/plugin_docs.py:
```diff
@@ -256,3 +262,34 @@ def get_versioned_doclink(path):
+                raise AnsibleError("Invalid format for section '%s': %s" % (s, to_native(e)))
+
+        if s in FRAGMENTS and isinstance(dstring, MutableMapping):
+            add_fragments(dstring, path, fragment_loader=floader, section=s)
```
This call is not passing `is_module`, which is required to know whether this is for a module or another plugin type.

bcoca replied:
not needed, since there is no difference in this case.

felixfontein replied:
It is needed, since doc fragments for plugins can also contain option deprecations.

felixfontein commented on lib/ansible/utils/plugin_docs.py:
```diff
@@ -208,18 +216,16 @@ def get_docstring(filename, fragment_loader, verbose=False, ignore_errors=False,
-            add_collection_to_versions_and_dates(data['returndocs'], collection_name, is_module=is_module, return_docs=True)
+            # add collection name to versions and dates
+            if collection_name is not None:
+                add_collection_to_versions_and_dates(data[key], collection_name, is_module=is_module)
```
This needs to pass `return_docs=True` for return docs.

felixfontein commented on lib/ansible/utils/plugin_docs.py:
```diff
@@ -179,17 +186,18 @@ def add_fragments(doc, filename, fragment_loader, is_module=False):
                     doc['seealso'] = []
                 doc['seealso'].extend(seealso)
 
-        if 'options' not in fragment:
```
The `add_collection_to_versions_and_dates` call in line 173 needs `return_docs=True` passed if this is a return value fragment.

bcoca committed:
ansible-doc add ability to use doc_fragments for RETURN

bcoca committed:
fix clog

bcoca committed:
added note to frag field

bcoca committed:
added merge

bcoca committed:
expand coverage

bcoca committed:
testit

bcoca committed:
added missing file

removed dastardly whitespace in comment that was about to destroy us all

felixfontein commented on lib/ansible/utils/plugin_docs.py:
```diff
@@ -170,33 +174,35 @@ def add_fragments(doc, filename, fragment_loader, is_module=False):
+                if seealso:
+                    if 'seealso' not in doc:
+                        doc['seealso'] = []
+                    doc['seealso'].extend(seealso)
```
While you're touching these lines, how about factoring out common code, similarly as `options` and `attributes` are handled together further below:
```suggestion
            for doc_key in ['notes', 'seealso']:
                if doc_key in fragment:
                    entries = fragment.pop(doc_key)
                    if entries:
                        if doc_key not in doc:
                            doc[doc_key] = []
                        doc[doc_key].extend(entries)
```

felixfontein commented on lib/ansible/utils/plugin_docs.py:
```diff
@@ -230,13 +236,16 @@ def get_docstring(filename, fragment_loader, verbose=False, ignore_errors=False,
             add_collection_to_versions_and_dates(data['returndocs'], collection_name, is_module=is_module, return_docs=True)
 
+        # add fragments to return
+        add_fragments(data['returndocs'], filename, fragment_loader=fragment_loader, is_module=is_module, section='RETURN')
```
I'm a bit worried that this will lead to sudden rejection of modules/plugins which have a return value called `extends_documentation_fragment` (while that would be quite surprising, it's technically allowed). How about doing something like:
```suggestion
        if not isinstance(data['returndocs'].get('extends_documentation_fragment'), Mapping):
            add_fragments(data['returndocs'], filename, fragment_loader=fragment_loader, is_module=is_module, section='RETURN')
```
Maybe with some warning/deprecation if `extends_documentation_fragment` turns out to be a mapping?

felixfontein commented on lib/ansible/utils/plugin_docs.py:
```diff
@@ -170,33 +174,35 @@ def add_fragments(doc, filename, fragment_loader, is_module=False):
         real_collection_name = '.'.join(real_fragment_name.split('.')[0:2]) if '.' in real_fragment_name else ''
         add_collection_to_versions_and_dates(fragment, real_collection_name, is_module=is_module)
```
```suggestion
        add_collection_to_versions_and_dates(fragment, real_collection_name, is_module=is_module, return_docs=(section == 'RETURN'))
```

bcoca committed:
Update lib/ansible/utils/plugin_docs.py

Co-authored-by: Felix Fontein <felix@fontein.de>

bcoca committed:
Apply suggestions from code review

Co-authored-by: Felix Fontein <felix@fontein.de>

bcoca committed:
remove mapping check

it should not conflict, if it did, we would have to check str/list but not mapping

felixfontein commented:
I created https://github.com/ansible/ansible/pull/85638 to fix the failing sanity tests in ansible-core devel.
